crates/datasources/src/sqlserver/mod.rs

```diff
@@ -0,0 +1,559 @@
+    macro_rules! make_column {
+        ($builder:ty, $rows:expr, $col_idx:expr) => {{
+            let mut arr = <$builder>::with_capacity($rows.len());
+            for row in $rows.iter() {
+                arr.append_option(row.try_get($col_idx)?);
+            }
+            Arc::new(arr.finish())
+        }};
+    }
+
+    let mut columns = Vec::with_capacity(schema.fields.len());
+    for (col_idx, field) in schema.fields.iter().enumerate() {
+        let col: Arc<dyn Array> = match field.data_type() {
+            DataType::Boolean => make_column!(BooleanBuilder, rows, col_idx),
+            DataType::Int16 => make_column!(Int16Builder, rows, col_idx),
+            DataType::Int32 => make_column!(Int32Builder, rows, col_idx),
+            DataType::Int64 => {
+                let mut arr = Int64Builder::with_capacity(rows.len());
+                for row in rows.iter() {
+                    let val: Option<Intn> = row.try_get(col_idx)?;
+                    arr.append_option(val.map(|v| v.0));
+                }
+                Arc::new(arr.finish())
+            }
+            DataType::Float32 => make_column!(Float32Builder, rows, col_idx),
+            DataType::Float64 => make_column!(Float64Builder, rows, col_idx),
+            DataType::Utf8 => {
+                // Assumes an average of 16 bytes per item.
+                let mut arr = StringBuilder::with_capacity(rows.len(), rows.len() * 16);
+                for row in rows.iter() {
+                    let val: Option<&str> = row.try_get(col_idx)?;
+                    arr.append_option(val);
+                }
+                Arc::new(arr.finish())
+            }
+            DataType::Binary => {
+                // Assumes an average of 16 bytes per item.
+                let mut arr = BinaryBuilder::with_capacity(rows.len(), rows.len() * 16);
+                for row in rows.iter() {
+                    let val: Option<&[u8]> = row.try_get(col_idx)?;
+                    arr.append_option(val);
+                }
+                Arc::new(arr.finish())
+            }
+            DataType::Timestamp(TimeUnit::Nanosecond, None) => {
+                let mut arr = TimestampNanosecondBuilder::with_capacity(rows.len());
+                for row in rows.iter() {
+                    let val: Option<NaiveDateTime> = row.try_get(col_idx)?;
+                    let val = val.map(|v| v.timestamp_nanos_opt().unwrap());
+                    arr.append_option(val);
+                }
+                Arc::new(arr.finish())
+            }
+
+            // TODO: All the others...
+            // Tiberius mapping: <https://docs.rs/tiberius/latest/tiberius/trait.FromSql.html>
+            other => {
+                return Err(SqlServerError::String(format!(
+                    "unsupported data type for sql server: {other}"
+                )))
+            }
+        };
+        columns.push(col);
+    }
```

Comment:
Possibly, it would require us to manage a thread pool. Also we'd probably want to do higher level parallelization through working with multiple partitions.

Reply:
I would think we would want both. The parallelization of multiple partitions would help with larger datasets that require several partitions, and would lend well to distributed execution. Without parallelizing the lower level stuff, we'll likely run into slower compute once we distribute the partitions as each partition would still be processing each column sequentially. 

this is definitely not blocking the PR approval, but just something to think about. _(it looks like our other sql readers follow the same pattern anyways)._

Reply:
https://github.com/GlareDB/glaredb/issues/2095